language_service/src/identifier_locator.rs

```diff
@@ -0,0 +1,327 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use std::mem::replace;
+use std::rc::Rc;
+
+use crate::qsc_utils::{
+    find_ident, resolve_item_relative_to_user_package, resolve_item_res, span_contains,
+    span_touches, Compilation,
+};
+use qsc::ast::visit::{walk_expr, walk_namespace, walk_pat, walk_ty_def, Visitor};
+use qsc::{ast, hir, resolve};
+
+#[allow(unused_variables)]
+pub(crate) trait LocatorAPI<'package> {
+    fn at_callable_def(
+        &mut self,
+        context: &LocatorContext<'package>,
+        decl: &'package ast::CallableDecl,
+    ) {
+    }
+
+    fn at_callable_ref(
+        &mut self,
+        path: &'package ast::Path,
+        item_id: &'_ hir::ItemId,
+        item: &'package hir::Item,
+        package: &'package hir::Package,
+        decl: &'package hir::CallableDecl,
+    ) {
+    }
+
+    fn at_new_type_def(&mut self, type_name: &'package ast::Ident, def: &'package ast::TyDef) {}
+
+    fn at_new_type_ref(
+        &mut self,
+        path: &'package ast::Path,
+        item_id: &'_ hir::ItemId,
+        item: &'package hir::Item,
+        package: &'package hir::Package,
+        type_name: &'package hir::Ident,
+        udt: &'package hir::ty::Udt,
+    ) {
+    }
+
+    fn at_field_def(
+        &mut self,
+        context: &LocatorContext<'package>,
+        field_name: &'package ast::Ident,
+        ty: &'package ast::Ty,
+    ) {
+    }
+
+    fn at_field_ref(
+        &mut self,
+        expr_id: &'package ast::NodeId,
+        field_ref: &'package ast::Ident,
+        item_id: &'_ hir::ItemId,
+        field_def: &'package hir::ty::UdtField,
+    ) {
+    }
+
+    fn at_local_def(
+        &mut self,
+        context: &LocatorContext<'package>,
+        pat: &'package ast::Pat,
+        ident: &'package ast::Ident,
+    ) {
+    }
+
+    fn at_local_ref(
+        &mut self,
+        context: &LocatorContext<'package>,
+        path: &'package ast::Path,
+        node_id: &'package ast::NodeId,
+        ident: &'package ast::Ident,
+    ) {
+    }
+}
+
+pub(crate) struct LocatorContext<'package> {
```

Comment:
To align with my `IdentVisitor` rename, maybe `IdentContext` for this? Or something else that makes sense to you.

Reply:
Discussed offline: the name makes more sense now with the renaming of `IdentifierLocator` to just `Locator`.